Chunks class iterator

FrancescElies opened the pull request:
Hi,

this PR is a proposal for a ```chunks class``` iterator, with two tests one for in-core and the other one for  out-of-core carray.
In case this changes are on the right path, would this fill the requirements to enable this kind of syntax ```for chunk_ in ca.chunks```?

Many thanks in advance for your input,

Regards,
F. Elies

esc commented:
From a first glance it looks really good. We probably also want some kind of typed version of this eventually, but this is the right direction for sure!

Two things I notice immediately:

a) Travis seems to be failing, haven't looked at it yet though.

b) You may want to update the changelog yourself as a third commit on this branch.

FrancescElies commented:
I broke something in the appending procedure, I will need some time to understand what's going on, I'll write back once I get what's happening, thanks!

FrancescElies commented:
As changelog entry do you mean a new entry in RELEASE_NOTES.rst?

esc commented:
Correct. I have been tending to having everyone write their changelog entries themselves, that way it saves me having to formulate and maintain them

FrancescElies commented:
Sure, please let me know if something is incorrect.

FrancescElies commented:
I have a small example that reproduces the strange behavior caught by the test suite.  
```python
In [1]: import numpy as np

In [2]: import bcolz

In [3]: a = np.arange(100)

In [4]: b = bcolz.carray(a, chunklen=10-1)

In [5]: b.append(a)

In [6]: for i in b.chunks:
   ...:         print i
   ...:
[0 1 2 3 4 5 6 7 8]
[ 9 10 11 12 13 14 15 16 17]
[18 19 20 21 22 23 24 25 26]
[27 28 29 30 31 32 33 34 35]
[36 37 38 39 40 41 42 43 44]
[45 46 47 48 49 50 51 52 53]
[54 55 56 57 58 59 60 61 62]
[63 64 65 66 67 68 69 70 71]
[72 73 74 75 76 77 78 79 80]
[81 82 83 84 85 86 87 88 89]
[90 91 92 93 94 95 96 97 98]
[99]
[99  0  1  2  3  4  5  6  7]
[ 8  9 10 11 12 13 14 15 16]
[17 18 19 20 21 22 23 24 25]
[26 27 28 29 30 31 32 33 34]
[35 36 37 38 39 40 41 42 43]
[44 45 46 47 48 49 50 51 52]
[53 54 55 56 57 58 59 60 61]
[62 63 64 65 66 67 68 69 70]
[71 72 73 74 75 76 77 78 79]
[80 81 82 83 84 85 86 87 88]
[89 90 91 92 93 94 95 96 97]
```

esc commented:
chunks should be uniform, I don't understand why there is a single chunk:

```
[99]
```

Also there should be some leftovers, namely:

```
[98 98]
```

esc commented on bcolz/carray_ext.pyx:
```diff
@@ -1216,9 +1229,17 @@ cdef class carray:
             self.chunks.append(chunk_)
             cbytes += chunk_.cbytes
         self.leftover = leftover = nbytes % self._chunksize
+
```
I think these changes break it? The tests all pass without it, so not sure why it would be needed.

FrancescElies replied:
yes I agree, something is going wrong with these changes, I am looking at it

FrancescElies replied:
I remembered why I added this, leftovers were not iterated for in memory carray, I updated the tests, but this is for sure what is causing the problems.
With this update on the tests https://github.com/FrancescElies/bcolz/commit/541e3abee475f94f13495b1481bc5b431b5d6d3c if you remove that code and rerun the suite you will see that in memory carray does not iterate over the leftovers.
```
FAIL: test00 (test_carray.chunkIterMemoryTest)
Testing chunk iterator including leftovers
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/francescelieshenar/01_working_dir/01_github/bcolz/bcolz/tests/test_carray.py", line 2164, in test00
    self.assertEquals(n, len(a) / chunklen_)
AssertionError: 9 != 10
```
PS.: you will see another failure, I don't know exactly why, but this is for sure the code causing problems

esc commented:
And yes, I presume it would work, for example in ``sum``:

```
diff --git c/bcolz/carray_ext.pyx i/bcolz/carray_ext.pyx
index f3f2b8d9dc..b073aff42b 100644
--- c/bcolz/carray_ext.pyx
+++ i/bcolz/carray_ext.pyx
@@ -1712,7 +1712,7 @@ cdef class carray:
 
         """
         cdef chunk chunk_
-        cdef npy_intp nchunk, nchunks
+        cdef npy_intp nchunks
         cdef object result
 
         if dtype is None:
@@ -1730,8 +1730,7 @@ cdef class carray:
         result = np.zeros(1, dtype=dtype)[0]
 
         nchunks = <npy_intp> cython.cdiv(self._nbytes, self._chunksize)
-        for nchunk from 0 <= nchunk < nchunks:
-            chunk_ = self.chunks[nchunk]
+        for chunk_ in self.chunks:
             if chunk_.isconstant:
                 result += chunk_.constant * self._chunklen
             elif self._dtype.type == np.bool_:
```

FrancescAlted commented on bcolz/carray_ext.pyx:
```diff
@@ -734,6 +734,19 @@ cdef class chunks(object):
+       self._iter_count = 0
+       return self
+
+    def __next__(self):
```
Hmm, I am not sure that adding an iterator directly in .chunks would be a good idea.  I think .chunks should be kept mostly hidden (in fact, maybe renaming it to ._chunks) would be good.  Regarding iterating chunks, would:

http://bcolz.blosc.org/reference.html#iterblocks

fill the need?

FrancescElies replied:
Maybe this function could do the trick, I'm not sure I never used it before, I'll check it, thanks for the tip!

FrancescElies replied:
I made some tests and I have a very good feeling about iterblocks, I would like to know what @CarstVaartjes and @esc think about this results and the future of this PR.

Please give me notice if this discussion should be moved to the mailing list.

In the following tests (v4 & v5 using iterblocks) (v2 @esc's code) (v1 & v3 much slower).
```python 
In [1]: from bcolz import carray

In [2]: from bcolz.carray_ext import test_v1, test_v2, test_v3, test_v4, test_v5

In [3]: import numpy as np

In [4]: n = 100000 + 3333

In [5]: x = np.arange(0, n, 1)

In [6]: c = carray(x, dtype='int64', chunklen=10000)

In [7]: %timeit test_v1(c)
1 loops, best of 3: 2.98 s per loop

In [8]: %timeit test_v2(c)
1000 loops, best of 3: 375 µs per loop

In [9]: %timeit test_v3(c)
1 loops, best of 3: 2.27 s per loop

In [10]: %timeit test_v4(c)
100 loops, best of 3: 4.1 ms per loop

In [11]: %timeit test_v5(c)
1000 loops, best of 3: 438 µs per loop
```
Code: https://github.com/FrancescElies/bcolz/blob/bench_iter_carray/bcolz/carray_ext.pyx#L2594-L2692

Hopefully next week I will be able to explore this a bit more.

esc commented:
I think the initial motivation for writing this iterator was to reduce code duplication inside the carray class. None the less I think that iterblocks might actually fullfill the requirement nicely. As a sidenote: I'm not sure that ``iterblocks`` is a good choice of name since ``blocks`` are already used inside blosc and are smaller units than ``chunks``.

Not sure exactly why it is slower, maybe because the Numpy array is allocated instead of being reused inside every loop iteration, but that is a guess.

FrancescElies commented:
I made the wrong assumption that  ```for chunk_ in ca.chunks``` should iterate the leftovers too, but your sum example made me reread the code and I saw that's not the way the design works (please correct me if I am misunderstanding it again), as you said it seems to work. 

If you guys decide this PR is still useful, I'll clean the git history and fix the tests.

FrancescElies commented:
Rebased, solved some conflicts, and refactored chunks as _chunks.
Hopefully rebasing is fine, in case you would prefer a new PR instead of rebasing I'll revert it.

esc commented:
Updating a pull-request with rebase and force-push is the mark of the git pro!

(i'll need some time to review it, though)

esc commented:
I only ever open new pull-requests if I want to add something someone elses Pull-Request. Github doesn't really support people working on the same branch if they all have their own dev repos, at least not cleanly.

FrancescElies commented:
About the time, of course, no problem! 
If somehow I could provide some help, please tell me

esc commented:
Do you still want this merged? If yes, please rebase, otherwise close it.

FrancescElies commented:
rebased (if in the end turns out to be completely unneeded it could be reverted )

esc commented:
As things have moved along now, I think that ``chunks`` shouldn't be renamed. It should be something that is visible to the programmer. In fact I have used this is in public examples so we shouldn't break this API.Can you remove bf409d7 and I'll get it merged up.

FrancescElies commented:
bang, bf409d7 gone

esc commented:
Danke u!

FrancescElies commented:
Kein Thema, thank u!
